Mathlib/Control/Monad/Cont.lean

```diff
@@ -120,6 +120,20 @@ instance (ε) [MonadExcept ε m] : MonadExcept ε (ContT r m) where
   throw e _ := throw e
   tryCatch act h f := tryCatch (act f) fun e => h e f
 
+instance (ε) [MonadExceptOf ε m] : MonadExceptOf ε (ContT r m) where
+  throw e _ := throw e
+  tryCatch act h f := tryCatch (act f) fun e => h e f
```

Comment:
Hmm, I know this was the pre-existing implementation of `MonadExcept`, but it's not good:
```
def foo : ContT Bool Lean.MetaM Bool := do
  let x ← try pure true catch _ => return false
  throwError "oh no {x}"

#eval foo.run pure
-- false
```
The exception from the continuation is able to throw back into the try/catch. I'll check the haskell library this module is based on to see if they have a solution.

Reply:
I don't think this is salvageable. The issue is that `act` needs to be constrained to `ContT r m r` so that we can delimit the continuation.
```
def ContT.tryCatch [Monad m] [MonadExceptOf ε m] (act : ContT r m r) (h : ε → ContT r m α) :
    ContT r m α := fun c =>
  MonadExceptOf.tryCatch (act pure) (fun e => h e c)
```

Reply:
By the way, we should add https://hub.darcs.net/ross/transformers/browse/Control/Monad/Trans/Cont.hs to the module docs since `mtl` imports `transformers`. There are still some definitions this library has that we don't have. (`evalContT` is what I called `ContT.delimit` elsewhere.)

Reply:
Maybe this is it? Notice that `ε` has to be in the same universe as `r`, and everything has to be constrained to `ContT r m r`.
```
def ContT.tryCatch (ε : Type u) [Monad m] [MonadExceptOf ε m]
    (act : ContT r m r) (h : ε → ContT r m r) :
    ContT r m r := fun c => do
  match ← try Sum.inr <$> act pure catch e => pure (Sum.inl e) with
    | Sum.inr v => c v
    | Sum.inl e => h e c
```

Reply:
> Hmm, I know this was the pre-existing implementation of MonadExcept

I've changed it to call the other implementation, so that we only have the incorrect code in one place rather than two!

Reply:
Maybe we should add my nonlocal `throwError` example into the docstring for `MonadExcept` as a warning?

Reply:
I was hoping something like this might work, but I guess there's nowhere to pass through `b`:
```
  tryCatch act h f := do
    let mut b : Bool := true
    try
      act (fun a => do
        try
          f a
        catch e2 =>
          b := false
          throw e2
      )
    catch e =>
      if b then h e f else throw e
```

Reply:
Yeah, I was thinking about wrapping the continuation in a try/catch too and came to the same conclusion.

In a way, try/catch for `ContT` isn't a good idea because you can't force a `finally` to actually run. You can always do an early exit from anywhere. In Scheme, they've worked out all this business with `dynamic-wind` to detect when you're calling a continuation and leaving a dynamic scope, so you can have continuation-safe `finally`s. It seems the only way to do this is to delimit continuations and chain them together (which is my vague understanding of how `dynamic-wind` works).

Reply:
> Maybe we should add my nonlocal `throwError` example into the docstring for `MonadExcept` as a warning?

Done.

Reply:
I made a [Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/MonadExcept.20in.20the.20ContT.20monad/near/375341221)